authz/providers/azure/azure.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright The Guard Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package azure
+
+import (
+	"strings"
+
+	"github.com/Azure/go-autorest/autorest/azure"
+	auth "github.com/appscode/guard/auth/providers/azure"
+	"github.com/appscode/guard/authz"
+	"github.com/appscode/guard/authz/providers/azure/data"
+	"github.com/appscode/guard/authz/providers/azure/rbac"
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+	authzv1 "k8s.io/api/authorization/v1"
+)
+
+const (
+	OrgType = "azure"
+)
+
+func init() {
+	authz.SupportedOrgs = append(authz.SupportedOrgs, OrgType)
+}
+
+type Authorizer struct {
+	rbacClient *rbac.AccessInfo
+}
+
+type authzInfo struct {
+	AADEndpoint string
+	ARMEndPoint string
+}
+
+func New(opts Options, authopts auth.Options, dataStore *data.DataStore) (authz.Interface, error) {
+	c := &Authorizer{}
+
+	authzInfoVal, err := getAuthzInfo(authopts.Environment)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error in getAuthzInfo %s")
+	}
+
+	switch opts.AuthzMode {
+	case ARCAuthzMode:
+		c.rbacClient, err = rbac.New(authopts.ClientID, authopts.ClientSecret, authopts.TenantID, authzInfoVal.AADEndpoint, authzInfoVal.ARMEndPoint, opts.AuthzMode, opts.ResourceId, opts.ARMCallLimit, dataStore, opts.SkipAuthzCheck)
+	case AKSAuthzMode:
+		c.rbacClient, err = rbac.NewWithAKS(opts.AKSAuthzURL, authopts.TenantID, authzInfoVal.ARMEndPoint, opts.AuthzMode, opts.ResourceId, opts.ARMCallLimit, dataStore)
+	}
+
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create ms rbac client")
+	}
+	return c, nil
+}
+
+func (s Authorizer) Check(request *authzv1.SubjectAccessReviewSpec) (*authzv1.SubjectAccessReviewStatus, error) {
+	if request == nil {
+		return nil, errors.New("subject access review is nil")
+	}
+
+	// check if user is service account
+	if strings.HasPrefix(strings.ToLower(request.User), "system") {
+		glog.V(3).Infof("returning no op to service accounts")
+		return &authzv1.SubjectAccessReviewStatus{Allowed: false, Reason: "no opinion"}, nil
+	}
+
+	// TODO: handle AKS glass break
```

Comment:
can we skip when oid is not defined in extra info? that should be pretty generic?

Reply:
should we skip or deny? This can be scenario with multiple auth?

Reply:
if oid is not defined, it means it's not AAD user. should be skip, I think.